File: DSA/src/Questions/String/TextJustification.java
package Questions.String;

import java.util.ArrayList;
import java.util.List;

/*
https://leetcode.com/problems/text-justification/description/
 Given an array of strings words and a width maxWidth, format the text such that each line has exactly maxWidth characters and is fully (left and right) justified.

You should pack your words in a greedy approach; that is, pack as many words as you can in each line. Pad extra spaces ' ' when necessary so that each line has exactly maxWidth characters.

Extra spaces between words should be distributed as evenly as possible. If the number of spaces on a line does not divide evenly between words, the empty slots on the left will be assigned more spaces than the slots on the right.

For the last line of text, it should be left-justified, and no extra space is inserted between words.

Note:

A word is defined as a character sequence consisting of non-space characters only.
Each word's length is guaranteed to be greater than 0 and not exceed maxWidth.
The input array words contains at least one word.
 

Example 1:

Input: words = ["This", "is", "an", "example", "of", "text", "justification."], maxWidth = 16
Output:
[
   "This    is    an",
   "example  of text",
   "justification.  "
]
Example 2:

Input: words = ["What","must","be","acknowledgment","shall","be"], maxWidth = 16
Output:
[
  "What   must   be",
  "acknowledgment  ",
  "shall be        "
]
Explanation: Note that the last line is "shall be    " instead of "shall     be", because the last line must be left-justified instead of fully-justified.
Note that the second line is also left-justified because it contains only one word.
Example 3:

Input: words = ["Science","is","what","we","understand","well","enough","to","explain","to","a","computer.","Art","is","everything","else","we","do"], maxWidth = 20
Output:
[
  "Science  is  what we",
  "understand      well",
  "enough to explain to",
  "a  computer.  Art is",
  "everything  else  we",
  "do                  "
]
 

Constraints:

1 <= words.length <= 300
1 <= words[i].length <= 20
words[i] consists of only English letters and symbols.
1 <= maxWidth <= 100
words[i].length <= maxWidth
Time Complexity: O(n)
Space Complexity: O(n)
 */
public class TextJustification {
    public List<String> fullJustify(String[] words, int maxWidth) {
        List<String> res = new ArrayList<>();
        List<String> curWords = new ArrayList<>();
        int curLen = 0;

        for (String word : words) {
            // If we can not fit in this line
            if (curLen + word.length() + curWords.size() > maxWidth) {
                // Total Spaces
                int totalSpaces = maxWidth - curLen;
                // Total Gaps
                int gaps = curWords.size() - 1;

                if (gaps == 0) {
                    // Only One word
                    res.add(curWords.get(0) + " ".repeat(totalSpaces));
                } else {
                    // Evenly distribute Spaces
                    int spacePerGap = totalSpaces / gaps;
                    // Add Extra Spaces from left
                    int extraSpaces = totalSpaces % gaps;

                    StringBuilder line = new StringBuilder();
                    for (int i = 0; i < curWords.size(); i++) {
                        line.append(curWords.get(i));
                        if (i < gaps) {
                            int spaceCount = spacePerGap + ((i < extraSpaces) ? 1 : 0);
                            line.append(" ".repeat(spaceCount));
                        }
                    }
                    res.add(line.toString());
                }
                curWords.clear();
                curLen = 0;
            }
            curWords.add(word);
            curLen += word.length();
        }

        StringBuilder lastLine = new StringBuilder(String.join(" ", curWords));
        while (lastLine.length() < maxWidth) {
            lastLine.append(' ');
        }
        res.add(lastLine.toString());

        return res;
    }

    public static void main(String[] args) {
        TextJustification textJustification = new TextJustification();
        String[] words = { "This", "is", "an", "example", "of", "text", "justification." };
        int maxWidth = 16;
        List<String> res = textJustification.fullJustify(words, maxWidth);
        for (String line : res) {
            System.out.println(line);
        }
    }
}

File: DSA/Python/Questions/String/TextJustification.py
"""
# Question: Text Justification
# Link: https://leetcode.com/problems/text-justification/

# Problem Statement:
# Given an array of strings words and a width maxWidth, format the text such that each line has exactly maxWidth characters 
# and is fully (left and right) justified.

# Example:
# Input: words = ["This", "is", "an", "example", "of", "text", "justification."], maxWidth = 16
# Output: [
#    "This    is    an",
#    "example  of text",
#    "justification.  "
# ]
"""

from typing import List


class TextJustification:
    def justify_text_standard(self, words: List[str], max_width: int) -> List[str]:
        """
        Standard approach using line-by-line justification
        Time Complexity: O(n) where n is total characters in all words
        Space Complexity: O(n) for storing result
        """
        result = []
        current_line = []
        current_length = 0

        i = 0
        while i < len(words):
            # Check if word can be added to current line
            if current_length + len(words[i]) + len(current_line) <= max_width:
                current_line.append(words[i])
                current_length += len(words[i])
                i += 1
            else:
                # Justify current line
                result.append(
                    self.justify_line(current_line, current_length, max_width, False)
                )
                current_line = []
                current_length = 0

        # Handle last line (left-justified)
        if current_line:
            result.append(
                self.justify_line(current_line, current_length, max_width, True)
            )

        return result

    def justify_line(
        self, line: List[str], line_length: int, max_width: int, is_last_line: bool
    ) -> str:
        """Helper function to justify a single line"""
        if len(line) == 1 or is_last_line:
            return " ".join(line).ljust(max_width)

        total_spaces = max_width - line_length
        gaps = len(line) - 1
        spaces_per_gap = total_spaces // gaps
        extra_spaces = total_spaces % gaps

        result = []
        for i in range(len(line) - 1):
            result.append(line[i])
            spaces = spaces_per_gap + (1 if i < extra_spaces else 0)
            result.append(" " * spaces)
        result.append(line[-1])

        return "".join(result)

    def justify_text_optimized(self, words: List[str], max_width: int) -> List[str]:
        """
        Optimized approach with minimal string operations
        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        result = []
        i = 0

        while i < len(words):
            # Calculate how many words fit on current line
            line_length = len(words[i])
            last_word = i

            while (
                last_word < len(words) - 1
                and line_length + len(words[last_word + 1]) + 1 <= max_width
            ):
                last_word += 1
                line_length += len(words[last_word]) + 1

            # Create the line with proper justification
            line = []
            word_count = last_word - i + 1

            if last_word == len(words) - 1 or word_count == 1:
                # Left justify last line or single word
                for j in range(i, last_word):
                    line.append(words[j])
                    line.append(" ")
                line.append(words[last_word])
                line.append(" " * (max_width - line_length))
            else:
                # Full justify middle lines
                spaces = max_width - (line_length - (word_count - 1))
                spaces_between = spaces // (word_count - 1)
                extra_spaces = spaces % (word_count - 1)

                for j in range(i, last_word):
                    line.append(words[j])
                    space_count = spaces_between + (1 if extra_spaces > 0 else 0)
                    line.append(" " * space_count)
                    extra_spaces -= 1
                line.append(words[last_word])

            result.append("".join(line))
            i = last_word + 1

        return result


def main():
    test_cases = [
        {
            "words": ["This", "is", "an", "example", "of", "text", "justification."],
            "max_width": 16,
            "expected": ["This    is    an", "example  of text", "justification.  "],
        },
        {
            "words": ["What", "must", "be", "acknowledgment", "shall", "be"],
            "max_width": 16,
            "expected": ["What   must   be", "acknowledgment  ", "shall be        "],
        },
    ]

    solution = TextJustification()
    for i, test in enumerate(test_cases):
        print(f"\nTest Case {i + 1}:")
        print(f"Words: {test['words']}")
        print(f"Max Width: {test['max_width']}")
        result_standard = solution.justify_text_standard(
            test["words"], test["max_width"]
        )
        result_optimized = solution.justify_text_optimized(
            test["words"], test["max_width"]
        )
        print("Standard Result:")
        for line in result_standard:
            print(f"'{line}'")
        print("Optimized Result:")
        for line in result_optimized:
            print(f"'{line}'")
        print("Expected:")
        for line in test["expected"]:
            print(f"'{line}'")
        assert (
            result_standard == result_optimized == test["expected"]
        ), f"Test case {i + 1} failed"


if __name__ == "__main__":
    main()
